### PR TITLE
163504614 create other API endpoints 

### PR DESCRIPTION
**Pull request by Jesse-efe**
#### What does this PR do?
Create more API endpoints

> GET api/v1//parties/:id/name

> PATCH api/v1/parties/:id

> DELETE api/v1/parties/:id

> POST api/v1/offices

> GET api/v1/offices

> GET api/v1/offices/:id

#### How should this be manually tested?
- clone this repo
- cd into it and run npm install
- run npm start
- use postman to test the above endpoints
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#163504493 #163504520 #163504538 #163504560 #163504592 #163504614
#### Screenshots
N/A
#### Questions:
N/A

### CODE REVIEW

File: source/controller/office.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import data from '../db/storage';
+
+class Office {
+  static async createOffice(req, res) {
```
Comment (akinwale-habibullah) [inline source/controller/office.js]: @Jesse-efe I'd like to ask these questions:

1. Why are you using async functions even when you are not performing any asynchronous task?

File: source/controller/office.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      id,
+      type,
+      name,
+    } ]
+    };
+
+    return res.status(201).json(response);
+  }
+
+  static async getAllOffice(req, res) {
```
Comment (akinwale-habibullah) [inline source/controller/office.js]: Same here. Its an async function even when no asynchronous task is being done.

File: source/controller/office.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    id = parseInt(id);
+    if (isNaN(id)) {
+      return res.status(400).json({
+        status: 400,
+        error: 'Invalid office Id'
+      });
+    }
+    let offices = data[1];
+    let foundOffice = false;
+    await offices.forEach((office) => {
```
Comment (akinwale-habibullah) [inline source/controller/office.js]: Any reason why you want to use await when iterating loop? It's not an asynchronous task.

**Review by akinwale-habibullah — COMMENTED** [review]
@Jesse-efe I'll like to ask this question:

Why are you using async functions even when you are not performing any asynchronous task? Are you thinking ahead and making plans for when you integrate a database? 

Some pieces of feedback I have for you:

- Put your routes in a directory instead of using a single module as you have now. Each resource should use a different route file in the routes directory.
- Remove the numerous res.status(XX).json you have in your validation.js. Use a better logic to check for an error flag as we discussed earlier. Keep track of what the user failed to pass in the request and return all of them. 
- For POST /Parties. In the middleware before the controller runs, ensure that logoUrl in the req matches a url format. Research on a regex pattern that will help you achieve this.

A commendation I have or you:

- I like how you handled potential exceptions/errors like you did in getOneParty, by using parseInt and checking if it is Nan. Kudos. But using async/await in these functions still looks odd (for now).

File: source/controller/party.js [hunk 4]
```diff
@@ -19,29 +19,129 @@ class Party {
...
+  }
+
+  static async editParty(req, res) {
+    const { id } = req.params;
+    const {
+      name,
+    } = req.body;
+    
+    let parties = data[0];
+    let foundParty = false;
```
Comment (akinwale-habibullah) [inline source/controller/party.js]: I think this variable isn't useful. If party.id === id at some point, you will return the response right? So if the function ever gets to line 103, you can go ahead to return res.status(XX).json.

What do you think?

File: source/helpers/validations.js [hunk 5]
```diff
@@ -1,51 +1,122 @@
...
+    return res.status(400).json({
+      status: 400,
+      error: 'party name was not specified',
+    });
+  }
+
+  req.body.name = name;
+  req.params.id = id;
+  next();
+}
```
Comment (akinwale-habibullah) [inline source/helpers/validations.js:83]: Add ;

**Review by akinwale-habibullah — APPROVED** [review]

**Comment by Jesse-efe** [comment]
Yes I plan making a database call at the points where I used await. Also, I feel the forEach loop would take a long time to run for large amount of data and would run asynchronously

**Comment by Jesse-efe** [comment]
I have also noted the feedback and would work on them

File: source/helpers/validations.js (hunk 5 above)
Comment (Jesse-efe) [reply source/helpers/validations.js:83]: Ok

**Pull request merged**